pkg/resourcevalidator/cache.go

```diff
@@ -0,0 +1,285 @@
+// Copyright 2019-2022 The Liqo Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package resourcevalidator
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"time"
+
+	"k8s.io/apimachinery/pkg/util/wait"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	sharing "github.com/liqotech/liqo/apis/sharing/v1alpha1"
+	vkv1alpha1 "github.com/liqotech/liqo/apis/virtualkubelet/v1alpha1"
+	"github.com/liqotech/liqo/pkg/discovery"
+)
+
+type peeringCache struct {
+	peeringInfo sync.Map
+	ready       bool
+}
+
+/**
+ * PeeringCache methods
+ */
+
+// Probe checks if the webhook cache is Ready.
+/* func Probe(req *http.Request) error {
+	if ready {
+		return nil
+	}
+	return fmt.Errorf("webhook cache not yet configured")
+} */
+
+func (pc *peeringCache) getPeeringFromCache(clusterID string) (*peeringInfo, bool) {
+	pi, found := pc.peeringInfo.Load(clusterID)
+	return pi.(*peeringInfo), found
+}
+
+func (pc *peeringCache) addPeeringToCache(clusterID string, pi *peeringInfo) {
+	pc.peeringInfo.Store(clusterID, pi)
+}
+
+func (pc *peeringCache) deletePeeringFromCache(clusterID string) {
+	pc.peeringInfo.Delete(clusterID)
+}
+
+// TODO: refresh has to be an update and not a new cache generation.
+func (spv *ShadowPodValidator) refreshCache(ctx context.Context) (done bool, err error) {
+	webhookrefreshlog.Info("[ REFRESH ] Refreshing peering cache")
+	c := spv.Client
+	shadowPodList := vkv1alpha1.ShadowPodList{}
+	resourceOfferList := sharing.ResourceOfferList{}
+	if err := c.List(ctx, &resourceOfferList, &client.ListOptions{}); err != nil {
+		return true, err
+	}
+	if !spv.PeeringCache.ready {
```

Comment:
In case the first iteration is completely different it is better to create two separate functions for increased clarity.